feature added to avoid joint limit with ComputeJointVelocityFromTwist

sjavdani opened the pull request:
Added a new objective function, to be used for ComputeJointVelocityFromTwist, which adds a cost and gradient to avoid joint limits. Also modified ComputeJointVelocityFromTwist in order to make it have a hard constraint near a limit (was previously hard constrained at 0, now constrained at whatever joint_limit_tolerance is)

siddhss5 commented on src/prpy/util.py:
```diff
@@ -428,11 +428,47 @@ def get_duration(self):
+    q_max = args[4]
+
+    #when joint is less than this distance away from limit, start penalizing
+    delta_j = 3e-1
```
Why is this hardcoded?

siddhss5 commented on src/prpy/util.py:
```diff
@@ -428,11 +428,47 @@ def get_duration(self):
+                 0. for diff_lower, diff_upper in zip(qdiff_lower, qdiff_upper)]
+
+    #constant for weighing cost of quadratic vs. being diff from bounds
+    lambda_dqdist = 0.01
```
Why is this hardcoded?

siddhss5 commented on src/prpy/util.py:
```diff
@@ -446,7 +482,8 @@ def quadraticObjective(dq, *args):
 
 def ComputeJointVelocityFromTwist(robot, twist,
-                                  objective=quadraticObjective,
+                                  objective=quadraticPlusJointLimitObjective,
```
I think `quadraticObjective` should be default in `prpy.util`. There are way too many parameters in `quadraticPlusJointLimitObjective` for this to be easily generalizable to all robots. If you feel like it, you should use the other objective with specially tuned parameters within a specific `robot.py`.

siddhss5 commented on src/prpy/util.py:
```diff
@@ -478,18 +515,20 @@ def ComputeJointVelocityFromTwist(robot, twist,
                                        bounds=dq_bounds, approx_grad=False)
 
     dq_opt = opt[0]
+#    if opt[1] > 0:
```
This has been removed in a downstream PR.

siddhss5 commented:
I'm concerned about this change. We use `VectorField` extensively and it's fast because `ComputeJointVelocityFromTwist` is fast. I'd be wary about adding more optimization logic. I understand stepwise performance can be better but I'm not sure it has been tested enough to warrant a wholesale switch.

mkoval commented:
I do agree that we should leave `quadraticObjective` as the default and move this to `prpy.util`. It does take a lot of parameters that may not have sane default values.

sjavdani commented:
Ok, changed the default to be quadratic objective. In order to make it so you can change the constants, I also changed it so some of the arguments had to be specified for both quadratic and quadratic+joint objectives.

Note that, compared to before, it has a different tolerance at which joint velocities are forced to be zero. I found this limit to work much better in practice - Sidd, if this messed up other stuff, feel free to make the constant (joint_limit_tolerance) smaller.

mkoval commented:
Those changes look good to me. :+1: 

@siddhss5 Any other comments? Or can I merge?
